book/chapter-01-unix-sockets/01_unix_sockets_intro.md

```diff
@@ -0,0 +1,159 @@
+### Streams
+
+Steams are also known as **connection based** sockets: this means that each peer in a connection (i.e. a server or client) is aware that a connection has been made, and that message has been received by the other party. With this one can send a continue _stream_ (of an arbritary length) of bytes. This is very useful where a guarantee that a message has been successfully sent: like an email for example.
+
+### Datagrams
+
+Datagrams are also known as **connection-less** sockets: i.e. the other party never knows if the other has left. A datagram is a discrete message with boundaries - there is a beginning, body, and end to each one.
+
+A nice analogy I have stolen from [this Stackoverflow response](https://stackoverflow.com/a/4688899) makes it a lot easier to visualise:
+
+> A stream socket is like a phone call -- one side places the call, the other answers, you say hello to each other (SYN/ACK in TCP), and then you exchange information. Once you are done, you say goodbye (FIN/ACK in TCP). If one side doesn't hear a goodbye, they will usually call the other back since this is an unexpected event; usually the client will reconnect to the server. There is a guarantee that data will not arrive in a different order than you sent it, and there is a reasonable guarantee that data will not be damaged.
+
+> A datagram socket is like passing a note in class. Consider the case where you are not directly next to the person you are passing the note to; the note will travel from person to person. It may not reach its destination, and it may be modified by the time it gets there. If you pass two notes to the same person, they may arrive in an order you didn't intend, since the route the notes take through the classroom may not be the same, one person might not pass a note as fast as another, etc.
+
+> So you use a stream socket when having information in order and intact is important. File transfer protocols are a good example here. You don't want to download some file with its contents randomly shuffled around and damaged!
+
+**Note well**: An important distinction to make with Unix datagram packets (compared to _network sockets_) is that the order of the packets are guaranteed, and the order will remain the same - but the analogy still holds!
+
+## An example
+We can create a quick example of how this works using python; although any language (including `ba/sh`) would be more than suffice here, although python has some advantages we can get into later (specifically - looking at contexts)!
+
+```python
+# server.py
+import socket
+
+with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
+    sock.bind("/tmp/docode.sock")
+    sock.listen()
+    print(f"Listening on {sock.getsockname()}")
+    conn, address = sock.accept()
+    with conn:
+        while True:
+            data = conn.recv(1024)
+            print(f"Got connection with message: {data}")
+            if not data:
+                break
+            conn.sendall(data)
+```
```

Comment:
This is nice and simple. The reader might want to run this to see it work in action. Can we setup a series of practical steps to run this? For instance, we can start with no `conn.sendall(data)` line initially and get the reader to run the python script and pass the data into the python process with the following bash script:

```bash
#!/bin/bash

# Start the Python server in the background
python3 server.py &

# Store the process ID
SERVER_PID=$!

# Give the server a moment to start up
sleep 1

# Send a message to the Unix socket
echo "Hello, Server" | socat - UNIX-CONNECT:/tmp/docode.sock

# Kill the Python server process
kill $SERVER_PID

# Clean up the socket file
rm /tmp/docode.sock

echo "Server stopped and socket file removed."

``` 
We could then write a python script that sends a message to the unix socket with another python script. We could also cover some of the following concepts:
-  `data = conn.recv(1024)`: the buffer and what this means, as in what if we receive a message that is bigger than the buffer? We will need a delimiter and a dynamic buffer that we can keep adding to until the delimiter is reached and that we know that the message is complete.
- connections: can we have multiple listeners? Can we have multiple publishers to the same unix socket? 
- cleaning up unix sockets in python and bash
All these are practical points that a reader would like to be guided on if they are to use this, otherwise they would walk away knowing about unix sockets, but will get frustrated hitting multiple roadblocks and having to google around to get something practical working with sockets

Reply:
Bufsizes ✅

Reply:
Cleaning up ✅